File: domusviewer/app/src/main/java/it/achdjian/paolo/domusviewer/DomusEngineRest/RequestAttributes.java
package it.achdjian.paolo.domusviewer.DomusEngineRest;

import android.content.SharedPreferences;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.util.Log;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.util.List;

import it.achdjian.paolo.domusviewer.DomusEngine;

/**
 * Created by Paolo Achdjian on 18/05/16.
 */
public class    RequestAttributes extends DomusEngineRest implements Stoppable {
    private DomusEngine.AttributesListener listener;
    private final Integer networkId;
    private final Integer endpointId;
    private final int clusterId;
    private final Integer[] attributes;
    private boolean toStop;

    public RequestAttributes(@NonNull SharedPreferences sharedPreferences, @NonNull ConnectionStatus connected, @NonNull DomusEngine.AttributesListener listener, @NonNull Integer networkId, @NonNull Integer endpointId, int clusterId, @NonNull Integer[] attributes) {
        super(sharedPreferences, connected);
        this.listener = listener;
        this.networkId = networkId;
        this.endpointId = endpointId;
        this.clusterId = clusterId;
        this.attributes = attributes;
        toStop = false;
    }

    @Override
    public void run() {
        StringBuilder builder = new StringBuilder("/devices/").
                append(Integer.toString(networkId,16)).
                append("/endpoint/").
                append(Integer.toString(endpointId,16)).
                append("/cluster/in/").
                append(Integer.toString(clusterId,16)).
                append("/attributes?id=").append(TextUtils.join(",", attributes));
        Log.d("REST",builder.toString());
        String body = getLongRead(builder.toString());
        if (body != null && toStop==false) {
            Log.d("REST respose", body);
            if (!body.isEmpty()) {
                ObjectMapper mapper = new ObjectMapper();
                try {
                    List<JSonAttribute> attributes = mapper.readValue(body, new TypeReference<List<JSonAttribute>>() {
                    });
                    listener.newAttributes(networkId, endpointId, attributes);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    @Override
    public void stop() {
        listener = null;
        toStop=true;
    }
}

File: cs5463/app/src/main/java/it/achdjian/paolo/cs5463/domusEngine/rest/RequestAttributes.kt
package it.achdjian.paolo.cs5463.domusEngine.rest

import android.util.Log
import com.fasterxml.jackson.core.type.TypeReference
import it.achdjian.paolo.cs5463.domusEngine.AttributeCoord
import it.achdjian.paolo.cs5463.domusEngine.DomusEngine
import it.achdjian.paolo.cs5463.domusEngine.MessageType
import java.io.IOException

/**
 * Created by Paolo Achdjian on 18/05/16.
 */
class RequestAttributes(val coord: AttributeCoord) : ZigbeeRunnable() {

    override fun run() {
        val builder = StringBuilder("/devices/").
                append(Integer.toString(coord.networkdId, 16)).
                append("/endpoint/").
                append(Integer.toString(coord.endpointId, 16)).
                append("/cluster/in/").
                append(Integer.toString(coord.clusterId, 16)).
                append("/attributes?id=").
                append(coord.attributeId.toString(16))
        Log.d(TAG, builder.toString())
        val body = DomusEngineRest.get(builder.toString())
        if (body.isNotBlank()) {
            try {
                val jsonAttributes = MAPPER.readValue<List<JSonAttribute>>(body, object : TypeReference<List<JSonAttribute>>() {})
                val attributes = Attributes(coord.networkdId, coord.endpointId, coord.clusterId, jsonAttributes)
                DomusEngine.handler.sendMessage(DomusEngine.handler.obtainMessage(MessageType.NEW_ATTRIBUTES, attributes))
            }catch (e: IOException) {
                Log.e(TAG, "Error parsing response for ${coord}", e)
                Log.e(TAG, "Response: " + body)
                e.printStackTrace()
            }
        }
    }
}
